jgrapht-core/src/main/java/org/jgrapht/util/DoublyLinkedList.java

```diff
@@ -17,473 +17,1363 @@
+    public void addNode(int index, ListNode<E> node)
+    {
+        if (index == size) { // also true if this is empty
+            linkLast(node);
+        } else {
+            ListNode<E> successor = index == 0 ? head : getNode(index);
+            linkBefore(node, successor);
+            if (head == successor) {
+                head = node;
+            }
+        }
     }
 
     /**
-     * Returns the number of elements in this list
-     *
-     * @return the number of elements in this list
+     * Inserts the specified {@link ListNode node} at the front of this list.
+     * <p>
+     * This method has constant runtime complexity O(1).
+     * </p>
+     * 
+     * @param node to add
+     * @throws IllegalArgumentException if {@code node} is already part of this or another
+     *         {@code DoublyLinkedList}
+     * @throws NullPointerException if {@code node} is {@code null}
      */
-    public int size() {
-        return size;
+    public void addNodeFirst(ListNode<E> node)
+    {
+        addNode(0, node);
     }
 
     /**
-     * Appends the specified value to the end of this list. This method
-     * returns the list node the {@code value} is contained in
-     *
-     * @param value the value to append to this list
-     * @return the list node, which contains the {@code value}
+     * Inserts the specified {@link ListNode node} at the end of this list.
+     * <p>
+     * This method has constant runtime complexity O(1).
+     * </p>
+     * 
+     * @param node to add
+     * @throws IllegalArgumentException if {@code node} is already part of this or another
+     *         {@code DoublyLinkedList}
+     * @throws NullPointerException if {@code node} is {@code null}
      */
-    public ListNode<E> add(E value) {
-        return addLast(value);
+    public void addNodeLast(ListNode<E> node)
+    {
+        addNode(size, node);
     }
 
     /**
-     * Inverts the list. For instance, calling this method on the list $(a,b,c,\dots,x,y,z)$
-     * will result in the list $(z,y,x,\dots,c,b,a)$. This method does only pointer manipulation,
-     * meaning that all the list nodes allocated for the previously added elements are valid after
-     * this method finishes.
+     * Inserts the specified {@link ListNode node} before the specified {@code successor} in this
+     * list.
+     * <p>
+     * This method has constant runtime complexity O(1).
+     * </p>
+     * 
+     * @param successor {@code ListNode} before which the {@code node} is inserted
+     * @param node to insert
+     * @throws IllegalArgumentException if {@code node} is already contained in this or another
+     *         {@code DoublyLinkedList} or {@code successor} is not contained in this list
+     * @throws NullPointerException if {@code successor} or {@code node} is {@code null}
      */
-    public void invert() {
-        if (size < 2) {
-            return;
+    public void addNodeBefore(ListNode<E> successor, ListNode<E> node)
```

Comment:
@Toptachamann what's your opinion about the parameter order for this method?
The successor is passed as first argument and the node as second which could be a little unintuitive.
But on the other hand, the add(int, E) method also has the index of the element in the list before the given element is inserted as first element. This consistency was my motivation in first place to make it this way.

Reply:
I would choose to have the parameters in the order they will be placed in the list, as this is intuitive. If we need to add the method `addNodeAfter(...)`, I would choose the option with `addNodeAfter( node, predecessor )`, since the first parameter should be the node we operate on.